File: 216. Combination Sum III.cpp
class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int> > res;
        vector<int> path;
        helper(k, n, path, res, 1);
        return res;
    }

    void helper(int cnt, int target, vector<int> path, vector<vector<int> > &res, int idx) {
        if (cnt == 0) {
            if (target == 0) {
                res.push_back(path);
            }
            return;
        } else if (target < 0 || idx > 9 || idx > target) {
            return;
        }
        for (int i = idx; i < 10; i++) {
            if (i > target) break;
            path.push_back(i);
            helper(cnt - 1, target - i, path, res, i + 1);
            path.pop_back();
        }
    }
};

File: java/216. Combination Sum III.java
// O(2^9)
class Solution {
    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(k, n, 1, new ArrayList<Integer>(), res);
        return res;
    }

    private void dfs(int k, int n, int start, ArrayList<Integer> path, List<List<Integer>> res) {
        if (k == 0) {
            if (n == 0) {
                res.add(new ArrayList<>(path));
            }
            return;
        }
        for (int i = start; i <= 9; ++i) {
            if (i <= n) {
                path.add(i);
                dfs(k - 1, n - i, i + 1, path, res);
                path.remove(path.size() - 1);
            }
        }
    }
}
